pkg/history/history.go

```diff
@@ -88,41 +88,77 @@ type ResponseHistory struct {
 }
 
 type History struct {
-	ProjectDir string
+	projectDir string
+	writeQueue chan func() error
+}
+
+func NewHistory(projectDir string) *History {
+	h := &History{
+		projectDir: projectDir,
+		writeQueue: make(chan func() error, 10), // Adjust the buffer size as needed
+	}
+
+	// Start the goroutine to handle write operations
+	go h.startWorker()
+
+	return h
+}
+
+func (h *History) startWorker() {
+	for writeFunc := range h.writeQueue {
+		if err := writeFunc(); err != nil {
+			fmt.Println("Error writing history file:", err)
+		}
+	}
 }
 
 func NewHistoryError(recordType RecordType, historyFile string) error {
 	return fmt.Errorf("could not write %s history to the JSON file '%s' due to a formatting issue. Please check the file's formatting and ensure it follows the correct JSON structure, or reset the history by deleting the file", recordType, historyFile)
 }
 
-func (h *History) WriteHistoryRecord(recordType RecordType, historyRecord *HistoryRecord) error {
-	historyFile, err := utils.NitricHistoryFile(h.ProjectDir, string(recordType))
+func (h *History) WriteHistoryRecord(recordType RecordType, historyRecord *HistoryRecord, successCallbacks ...func()) error {
+	historyFile, err := utils.NitricHistoryFile(h.projectDir, string(recordType))
 	if err != nil {
 		return err
 	}
 
-	existingRecords, err := h.ReadHistoryRecords(recordType)
-	if err != nil {
-		return NewHistoryError(recordType, historyFile)
+	// Define the write operation as a function
+	writeOp := func() error {
+		existingRecords, err := h.ReadHistoryRecords(recordType)
+		if err != nil {
+			return NewHistoryError(recordType, historyFile)
+		}
+
+		existingRecords = append(existingRecords, historyRecord)
+
+		data, err := json.Marshal(existingRecords)
+		if err != nil {
+			return NewHistoryError(recordType, historyFile)
+		}
+
+		err = os.WriteFile(historyFile, data, fs.ModePerm)
+		if err != nil {
+			return NewHistoryError(recordType, historyFile)
+		}
+
+		// Call each success callback function after successful write
+		for _, callback := range successCallbacks {
+			if callback != nil {
+				callback()
+			}
+		}
+
+		return nil
 	}
 
-	existingRecords = append(existingRecords, historyRecord)
-
-	data, err := json.Marshal(existingRecords)
-	if err != nil {
-		return NewHistoryError(recordType, historyFile)
-	}
-
-	err = os.WriteFile(historyFile, data, fs.ModePerm)
-	if err != nil {
-		return NewHistoryError(recordType, historyFile)
-	}
+	// Enqueue the write operation
+	h.writeQueue <- writeOp
 
 	return nil
 }
```

Comment:
Does the channel need to accept callbacks which have their own slices of callbacks? Can the history writer just accept history entries as structs? And the function that's queuing it can treat it as successful once queued?

Reply:
See new updates